integrationtests/testground/main.go

```diff
@@ -0,0 +1,453 @@
+		env.RecordMessage(msg, args...)
+	}
+	debug = func(msg string, args ...interface{}) {
+		if env.IntParam("verbose") > 1 {
+			env.RecordMessage(msg, args...)
+		}
+	}
+	fail = func(msg string, args ...interface{}) {
+		env.RecordFailure(fmt.Errorf(msg, args...))
+	}
+	setup(env, ic)
+	for i, pair := range simulatedNetworks {
+		round := pair.simulation
+		if ic.GlobalSeq == 1 {
+			msg("################### Round %s network ###################", round)
+			ic.NetClient.MustConfigureNetwork(context.Background(), &network.Config{
+				Network:       "default",
+				Enable:        true,
+				Default:       pair.shape,
+				CallbackState: sync.State("network-configured-" + round),
+				RoutingPolicy: network.AllowAll,
+			})
+			msg("Done configuring %s network", round)
+		}
+
+		desiredAddr := ""
+		if env.TestSidecar {
+			// listen on non-local interface when running in container, so the peers can communicate with each other
+			ip := ic.NetClient.MustGetDataNetworkIP().String()
+			desiredAddr = fmt.Sprintf("/ip4/%s/tcp/%d", ip, 3000+i)
+		}
+		if err := testRound(env, ic, round, desiredAddr); err != nil {
+			msg("################### Peer #%d with %s network failed: %v ###################", ic.GlobalSeq, round, err)
+
+			return err
+		}
+	}
+	return nil
+}
+
+func testRound(env *runtime.RunEnv, ic *run.InitContext, round string, desiredAddr string) error {
+	ctx, _ := context.WithTimeout(context.Background(), time.Minute)
+	var cli *client.Client
+	var stop func()
+	if !env.BooleanParam("late-start") {
+		var err error
+		cli, stop, err = startClient(desiredAddr, env, ic)
+		if err != nil {
+			return err
+		}
+	}
+
+	chThreadToJoin := make(chan *SharedInfo, 1)
+	topic := sync.NewTopic("thread-"+round, &SharedInfo{})
+	// choose a single peer to create the thread and broadcast, then each peer (include this one itself) creates its own log on this thread.
+	if !env.BooleanParam("late-start") && !env.BooleanParam("early-stop") && ic.GroupSeq == 1 {
+		thr, err := createThread(ctx, cli)
+		if err != nil {
+			msg("Failed to create the thread: %v", err)
+			return err
+		}
+		msg("Created thread")
+		ic.SyncClient.MustPublishSubscribe(ctx,
+			topic,
+			thr.Sharable(),
+			chThreadToJoin)
+		msg("Published thread")
+	} else {
+		ic.SyncClient.MustSubscribe(ctx, topic, chThreadToJoin)
+	}
+
+	// peers each send their own number and have a consensus on the correct total
+	publishAndGetTotal := func(topic *sync.Topic, myNum int) (total int) {
+		ch := make(chan int)
+		ic.SyncClient.MustPublishSubscribe(ctx, topic, myNum, ch)
+		for i := 0; i < env.TestInstanceCount; i++ {
+			total += <-ch
+		}
+		return
+	}
+	myRecords := env.IntParam("records")
+	meLive := 1
+	if env.BooleanParam("late-start") {
+		myRecords = 0
+		meLive = 0
+	}
+	expectedRecords := publishAndGetTotal(sync.NewTopic("expectedRecords-"+round+"-step-1", 0), myRecords)
+	livePeers := publishAndGetTotal(sync.NewTopic("livePeers-"+round+"-step-1", 0), meLive)
```

Comment:
Each instance runs independently, and there's no way for an instance to get the full picture of how many instances in each group, so the instances need to communicate with each other to get the correct number.

Reply:
So this is some kind of pubsub between instances?

Reply:
It's more like `atomic.AddInt64()` across processes, implemented by Testground via a central Redis.